src/main/java/com/microsoft/jenkins/azuread/AzureAdUser.java

```diff
@@ -68,13 +71,15 @@ public static AzureAdUser createFromJwt(JwtClaims claims) {
         return user;
     }
 
-    public void setAuthorities(Collection<String> groups) {
-        GrantedAuthority[] newAuthorities = new GrantedAuthority[groups.size() + 1];
+    public void setAuthorities(Collection<ActiveDirectoryGroup> groups) {
+        GrantedAuthority[] newAuthorities = new GrantedAuthority[groups.size() * 2 + 2];
         int i = 0;
-        for (String objectId : groups) {
-            newAuthorities[i++] = new AzureAdGroup(objectId);
+        for (ActiveDirectoryGroup group : groups) {
+            newAuthorities[i++] = new AzureAdGroup(group.id(), group.name());
+            newAuthorities[i++] = new GrantedAuthorityImpl(group.id());
         }
-        newAuthorities[i] = SecurityRealm.AUTHENTICATED_AUTHORITY;
+        newAuthorities[i++] = SecurityRealm.AUTHENTICATED_AUTHORITY;
+        newAuthorities[i] = new GrantedAuthorityImpl(objectID);
```

Comment:
This is for backwards compatibility when someone has a configuration where objectID of user is used. With this line we make sure his/her configuration still works

Reply:
As now we use userUniqueName as the principal

Reply:
Line 79 will add all objectId as authenticated.

Reply:
It will only add objectIds of the groups not the current user.

Reply:
Current user should not be authenticated by default, otherwise anyone can login into the jenkins server.

Reply:
You get the current user once the login is done, isn't it? I am not sure how anyone can login to jenkins server with this change?
This is to support the permissions based on both objectID and unique name of user. When we add objectIds and names of all the groups user belong to, we add object id of user as well.

Reply:
Once the login is done, it is only authenticated with AAD. Not every AAD user will be a Jenkins user.

Reply:
Sorry mate, I dont follow.
If you think this is not needed. Please let me know the other way of supporting configurations where roles are assigned to per user not group before upgrade. Once we upgrade we start using user unique name as principal instead of id and it will break the setup

Reply:
I am just a little confused here. `newAuthorities[i++] = new GrantedAuthorityImpl(group.id());` should replace the origin `newAuthorities[i++] = new AzureAdGroup(objectId);`, and `newAuthorities[i++] = new AzureAdGroup(group.id(), group.name());` will enable use unique name to authenticate. Then why do we need a new line `newAuthorities[i] = new GrantedAuthorityImpl(objectID);`? Why this is not covered in the above two lines?

Reply:
As far as my understanding is, every object in Azure AD is an object with id, be it a user or group.

```
        for (ActiveDirectoryGroup group : groups) {
            newAuthorities[i++] = new AzureAdGroup(objectId);	            newAuthorities[i++] = new AzureAdGroup(group.id(), group.name());
            newAuthorities[i++] = new GrantedAuthorityImpl(group.id());
        }	        }
```
These lines adds names and ids of groups the current user belongs to authorities which Jenkins can use

```
newAuthorities[i] = new GrantedAuthorityImpl(objectID);
```
This line adds the object id of the user to authorities. This line enables object id of user to be used in permissions and config. 
Consider the following scenario ( Before this patch )
I have a jenkins instance with Role based authentication, i can assign roles using object ids. So I assign roles **Developer** to all the users part of **DL-Developers** group using the object id of DL-Developers group. 
I can also assign role **admin** to specific users using their object ids. This works.

After this patch, specifically `AzureAuthenticationToken.java` changes the `getPrincipal()` to return `azureAdUser.getUniqueName()` instead of `azureAdUser.getObjectID()`. So after this patch Jenkins wont authorise user based on his object ID. 
To be able to let Jenkins know the object ID of user we add it to authorities and Jenkins will be able to provide admin permissions to user.

```
newAuthorities[i] = new GrantedAuthorityImpl(objectID);
```

Reply:
Thanks @shantur. It makes senses to me.

Reply:
@gavinfish : Brilliant, I will resolve this then.